pandas/_libs/lib.pyx

```diff
@@ -2357,3 +2354,190 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
+                                                 Py_ssize_t rows_count,
+                                                 Py_ssize_t col_count,
+                                                 bint keep_trivial_numbers):
+    """
+    Concatenates `rows_count` elements from each `col_count` numpy arrays
+    in `date_cols` into strings.
+
+    Note
+    ----
+    This function speeds up concatenation for numpy arrays.
+    You also can use `_concat_date_cols_sequence` function.
+
+    Parameters
+    ----------
+    date_cols : tuple of numpy arrays
+    rows_count : Py_ssize_t
+    col_count : Py_ssize_t
+    keep_trivial_numbers : bool, default False
+        If True, then for the case of one sequence in `date_cols`,
+        conversion (to string from integer/float zero) is not performed
+
+    Returns
+    -------
+    arr_of_rows : 1-d numpy array
+    """
+    cdef:
+        Py_ssize_t col_idx, row_idx
+        list list_to_join
+        cnp.ndarray[object] iters
+        object[::1] iters_view
+        flatiter it
+        cnp.ndarray[object] result
+        object[:] result_view
+
+    result = np.zeros(rows_count, dtype=object)
+    result_view = result
+
+    if col_count == 1:
+        array = date_cols[0]
+        it = <flatiter>PyArray_IterNew(array)
+        for row_idx in range(rows_count):
+            item = PyArray_GETITEM(array, PyArray_ITER_DATA(it))
+            convert_and_set_item(item, row_idx, result_view,
+                                 keep_trivial_numbers)
+            PyArray_ITER_NEXT(it)
+    else:
+        list_to_join = [None] * col_count
+        # setup iterators
+        iters = np.zeros(col_count, dtype=object)
+        iters_view = iters
+        for col_idx, array in enumerate(date_cols):
+            iters_view[col_idx] = PyArray_IterNew(array)
+        for row_idx in range(rows_count):
+            for col_idx, array in enumerate(date_cols):
+                it = <flatiter>iters_view[col_idx]
+                item = PyArray_GETITEM(array, PyArray_ITER_DATA(it))
+                put_object_as_unicode(list_to_join, col_idx, item)
+                PyArray_ITER_NEXT(it)
+            result_view[row_idx] = PyUnicode_Join(' ', list_to_join)
+
+    return result
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef cnp.ndarray[object] _concat_date_cols_sequence(tuple date_cols,
+                                                    Py_ssize_t rows_count,
+                                                    Py_ssize_t col_count,
+                                                    bint keep_trivial_numbers):
+    """
+    Concatenates `rows_count` elements from each `col_count` sequences
+    in `date_cols` into strings.
+
+    Parameters
+    ----------
+    date_cols : tuple of sequences
+    rows_count : Py_ssize_t
+    col_count : Py_ssize_t
+    keep_trivial_numbers : bool, default False
+        If True, then for the case of one sequence in `date_cols`,
```

Comment:
I have no idea what this means

Reply:
This optimization is performed if ` keep_trivial_numbers == True and len(date_cols) == 1`